server/src/main/java/org/opensearch/index/shard/CloseableRetryableRefreshListener.java

```diff
@@ -80,64 +99,86 @@ private void scheduleRetry(TimeValue interval, String retryThreadPoolName, boole
             || retryScheduled.compareAndSet(false, true) == false) {
             return;
         }
+
         boolean scheduled = false;
         try {
-            this.threadPool.schedule(() -> {
-                boolean successful;
-                boolean permitAcquired = semaphore.tryAcquire();
-                try {
-                    successful = permitAcquired && performAfterRefresh(didRefresh, isRetry);
-                } finally {
-                    if (permitAcquired) {
-                        semaphore.release();
-                    }
-                    retryScheduled.set(false);
-                }
-                scheduleRetry(successful, didRefresh, isRetry || permitAcquired);
-            }, interval, retryThreadPoolName);
+            this.threadPool.schedule(
+                () -> runAfterRefreshWithPermit(didRefresh, () -> retryScheduled.set(false)),
+                interval,
+                retryThreadPoolName
+            );
             scheduled = true;
-            getLogger().info("Scheduled retry with didRefresh={} isRetry={}", didRefresh, isRetry);
+            getLogger().info("Scheduled retry with didRefresh={}", didRefresh);
         } finally {
             if (scheduled == false) {
                 retryScheduled.set(false);
             }
         }
     }
 
+    /**
+     * Runs the performAfterRefresh method under permit. If there are no permits available, then it is no-op. It also hits
+     * the scheduleRetry method with the result value of the performAfterRefresh method invocation.
+     */
+    private void runAfterRefreshWithPermit(boolean didRefresh, Runnable runFinally) {
+        boolean successful;
+        boolean permitAcquired = semaphore.tryAcquire();
+        try {
+            successful = permitAcquired && performAfterRefresh(didRefresh);
+        } finally {
+            if (permitAcquired) {
+                semaphore.release();
+            }
+            runFinally.run();
+        }
+        assert permitAcquired;
```

Comment:
Is it guaranteed to have the permit at this step?

Reply:
No, it is not guaranteed. There can be cases where after the closed check and before acquiring a permit, the close method invocation has acquired all the permits, in which case this would lead to permitAcquired being false. It is difficult to repro this in test.

Reply:
Should we have a timeout to try acquire?

Reply:
Right now, it would return false if the permits were unavailable. However, we have retry being scheduled to try the execution again. If the underlying refresh listener has been closed (which means all permits have been acquired), then we would not be retrying as well based on the atomic boolean closed flag.